packages/scripts/src/findFaq.ts

```diff
@@ -166,3 +174,127 @@ export const findFaq = async ({
+        filter,
+        numCandidates,
+      }: Partial<FindNearestNeighborsOptions> = {
+        // Default options
+        indexName: "vector_index",
+        path: "embedding",
+        k: 10,
+        minScore: 0.95,
+        // User options override
+        ...(options ?? {}),
+      };
+      return collection
+        .aggregate<WithScore<WithId<FaqEntry>>>([
+          {
+            $vectorSearch: {
+              index: indexName,
+              queryVector: vector,
+              path,
+              limit: k,
+              numCandidates: numCandidates ?? k * 15,
+              filter,
+            },
+          },
+          {
+            $addFields: {
+              score: {
+                $meta: "vectorSearchScore",
+              },
+            },
+          },
+          {
+            $match: {
+              score: { $gte: minScore },
+            },
+          },
+        ])
+        .toArray();
+    },
+  };
+};
+
+/**
+  For each given question, finds if any similar messages already have a faqId.
+  If so, adopts the faqId. Otherwise, invents a new faqId for this category of
+  question.
+ */
+export const assignFaqIds = async ({
+  faq,
+  faqCollection,
+  backportNewIds,
+}: {
+  faq: FaqEntry[];
+  faqCollection: Collection<
+    WithId<FaqEntry & { created: Date; epsilon: number }>
+  >;
+
+  /**
+    If a new faqId is generated because existing entries did not have a faqId,
+    assigns the new faqId to the existing entries. 
+   */
+  backportNewIds?: boolean;
+}): Promise<(FaqEntry & { faqId: string })[]> => {
+  const store = makeFaqVectorStoreCollectionWrapper(faqCollection);
+
+  return await Promise.all(
+    faq.map(async (q) => {
+      // See if there already is an ID for this FAQ.
+      const previousFaqs = await store.findNearestNeighbors(q.embedding);
+      const previousFaqsWithFaqIds = previousFaqs.filter(
+        (q) => q.faqId !== undefined
+      );
+      previousFaqsWithFaqIds.sort((a, b) => b.score - a.score);
+
+      // Use the pre-existing faqId or generate a new one for this category
+      const faqId =
+        previousFaqsWithFaqIds[0]?.faqId ?? ObjectId.generate().toString("hex");
+      console.log(
+        `${
+          previousFaqsWithFaqIds[0]?.faqId === undefined
+            ? "Generated new"
+            : "Reused existing"
+        } faqId ${faqId} for question category "${q.question}"`
+      );
+
+      if (backportNewIds) {
+        const previousFaqsWithoutIds = previousFaqs.filter(
+          (q) => q.faqId === undefined
+        );
+        await Promise.all(
+          previousFaqsWithoutIds.map(async (q) => {
+            const updateResult = await faqCollection.updateOne(
+              {
+                _id: q._id,
+              },
+              { $set: { faqId } }
+            );
+            console.log(
+              `Backported new ID ${faqId} to ${q._id} (modified ${updateResult.modifiedCount})`
+            );
+          })
+        );
```

Comment:
This might be a good opportunity for `bulkWrite` since this could fire off up to `k` db calls per FAQ cluster. Probably not a big deal at our scale but something to consider
```suggestion
        await faqCollection.bulkWrite(
          previousFaqsWithoutIds.map((q) => ({
            updateOne: {
              filter: { _id: q._id },
              update: { $set: { faqId } },
            },
          }))
        );
```